File: small_problems/medium_1/diamonds.rb
=begin
PEDAC
  start 826
  finish 8:44
  time 18min
Problem
  rules
    accept an odd integer as an argument `n`
    print a diamond using the * character across `n` lines where
      the first line starts with 1 center aligned *
      each line increases in size by 1 star, still centered
      after reaching `n` * characters, each line decreases by 1
  clarifying questions?
    reject a negative or zero input? or output nothing?
      assume will be greater than 0
Algorithm
  make array of star counts per line
    from 1 up to n by 2
      [1, 3, 5, 7, 9]
    from n-1 down to 1 by 2 (slice and reverse prior array)
      [7, 5, 3, 1]
  for each line
    print a center aligned star with padding based on n
=end

def diamond(n)
  lines = 1.step(by: 2, to: n).to_a
  lines.push(*lines[0..-2].reverse)

  lines.each { |stars| puts ('*' * stars).center(n) }
end

# diamond(9)
diamond(5)

File: js_small_problems/interpretive_problem_solving/diamonds.js
/*
start 9:54
finish 10:26
takeaways
*/

/*
Problem
  input: odd integer
  output: n x n grid of stars in the shape of a diamond, logged to console
  rules:
    accept an odd integer as input 'n'
    log a grid of size n x n
    place '*' chars in the grid
    first row should have 1 star
    increase the number of stars on each row by 2 until reaching n
    then decrease by 2 until reaching 1
    pad each row with spaces to reach the correct number of characters for each row
  questions
    - should i assume the input will always be a Number type? yes
      - how should i handle non Number types that can be converted to numbers? n/a
    - will the input ever be 0 or negative? how should i handle those cases? possible
    - how should i handle missing or extra inputs? assume not applicable
Examples
  expected inputs:
    number type odd integer
  malformed inputs
    negative value? -1
    zero value? 0
Data Structure
  input: integer
  intermediate:
    string for each row
    array to capture each row of the grid
  output: n/a
Algorithm
  if n is less than 1, return nothing
  initialize a counter to 1
    5
  initialize stars to 1
  initialize rowLength to n
  iterate up to counter
    log stars, padded with rowLength - stars spaces
      n === 5
        stars === 1
        rowLength === 5
        padding === 5-1
        each side gets 2 spaces
    if counter is > ceil(n / 2) stars - 2
    else stars + 2
*/
function diamond(n) {
  if (n < 1) return;

  let stars = 1;
  let rowLength = n;

  for (let counter = 1; counter <= n; counter += 1) {
    let padding = ' '.repeat((rowLength - stars) / 2)
    console.log(padding + '*'.repeat(stars) + padding);

    stars += counter < Math.ceil(n / 2) ? 2 : -2
  }
}


// diamond(-1);
// diamond(0);

// diamond(1);
// // logs
// // *

diamond(3);
// // logs
// // *
// // ***
// // *
diamond(5);
diamond(9);
// // logs
// //     *
// //   ***
// //   *****
// // *******
// // *********
// // *******
// //   *****
// //   ***
// //     *